guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+            response.set_header("Content-Type", "application/json");
+            Ok(response)
+        })
+    }
+}
+```
+
+Finally the `Handler` passed to `Server::run` must use `Request` and `Response`:
+
+```rust
+impl Server {
+    async fn run<T>(self, mut handler: T) -> Result<(), Error>
+    where
+        T: Handler<Request, Response = Response>,
+    {
+        // ...
+    }
+}
+```
+
+Creating the server has not changed:
+
+```rust
+let handler = RequestHandler;
+let handler = Timeout::new(handler, Duration::from_secs(30));
+let handler = JsonContentType::new(handler);
+
+server.run(handler).await
+```
+
+So, we now have a `Handler` trait that makes it possible to break our
+application up into small independent parts and re-use them. Not bad!
+
+## "What if I told you..."
+
+Until now we've only talked about HTTP from the servers perspective. But our
+`Handler` trait actually fits HTTP clients as well. One can imagine a client
+`Handler` that accepts some `Request` struct and asynchronously sends it to
+someone on the internet. Our `Timeout` wrapper is actually useful here as well.
+`JsonContentType` probably isn't since its not the clients job to set response
+headers.
+
+Since our `Handler` trait is useful for defining both servers and clients,
+`Handler` probably isn't an appropriate name anymore. A client doesn't handle an
+request, it sends it to a server and the server handles it. Lets instead call
+our trait `Service`:
+
+```rust
+trait Service<Request> {
+    type Response;
+    type Error;
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    fn call(&mut self, request: Request) -> Self::Future;
+}
+```
+
+This is actually _almost_ the `Service` trait defined in Tower. If you've been
+able to follow along until this point you now understand most of Tower. Besides
+the `Service` trait Tower also provides several utilities that implement
+`Service` by wrapping some other type that also implements `Service`. Exactly
+like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
+similar to what we've just did.
+
+Some example services provided by Tower:
+
+- `Timeout` - This is pretty much identical to the timeout we have built.
+- `Retry` - To automatically retry failed requests.
+- `RateLimit` - To limit the number of requests a service can handle over a
+  period of time.
+
+Types like `Timeout` and `JsonContentType` are typically called "middlewares"
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called "leaf services"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
```

Comment:
Might wanna link https://aws.amazon.com/builders-library/using-load-shedding-to-avoid-overload/ and there are a few other articles in there that are REALLY good and connect the real world with these concepts. Maybe for later in this section.

Reply:
Yeah, there are a few articles in that builder library that are great reads and connect some real world concepts. Load shedding tends to be the root behavior where you reject requests when the buffer is full and that rejection stops requests from coming in etc. What ever you feel is correct :)

Reply:
I actually thinking linking to both is fine 😊